take_along_axis with non fusion inputs

naoyam opened the pull request:
This PR extends the scheduler support of `take_along_axis` so hat its input tensor no longer needs to be a fusion input. Much like `pad` and `slice`, when there's a mismatched parallelization between the input and output tensors of a `take_along_axis` op, the input tensor is automatically placed in a shared memory, which may be `MemoryType::Shared` or `MemoryType::Global` depending on how the op is parallelized.

The current scheduling policy is to make `take_along_axis` handled as if it were a pointwise op, even though it isn't. It may not be optimal, but I don't think we want to have another scheduler just for these gather-like ops, but rather these ops should be fused with reductions and normalizations if possible, so I think it makes most sense to extend the existing scheduler to be able to fuse `take_along_axis`. 

To make `take_along_axis` schedulable like normal pointwise ops, I extended the transform replay and parallelization propagation so that propagations are permissively done across index-accessed domains. For example, when we have:

```
auto tv0 = [I0, I1];
auto tv1 = take_along_axis(tv0, index_tensor, 1);
```

And:

```
tv1->split(1, 4);
```

Then, `replayPasC` with `tv0` as a producer and `tv1` as a consumer would replay the split of `tv1` on `tv0->axis(1)`.

This permissive propagation would be fine if the input and output domains have similar sizes, but otherwise, it may be more efficient to use different transformations. That is not done in this PR, but rather the focus is to make these ops be fused together as much as possible. 

The decision of the validity of fusing these gather-like ops mostly just depends on whether all IterDomains are mapped with scheduler reference tensors. For example, in a pointwise case like below:

```
auto tv0 = [I0, I1];
auto tv1 = take_along_axis(tv0, index_tensor, 1);
fusion.addOutput(tv1);
```

`tv1` would be picked as the reference tensor of scheduling, and the scheduler analyzes the fusion if all input tensors are mapped with the reference. This mapping only meant the exact mapping, so it would fail for the domain of `I1`. It is now extended to allow mappings through index-based accesses. 

In the case of the reduction and normalization schedulers, the output tensor of a reduction expr is used as the reference. When `take_along_axis` is used after the reduction, we should be able to schedule it just like any other pointwise ops. When `take_along_axis` is used before the reduction, if the indexed domain becomes a size-1 domain and is squeezed, we can no longer propagate anything from the reduction tensor to the indexed domain, so in that case the reduction scheduler needs to reject the fusion. It is also the same in the normalization scheduler.

This can be similarly extended for `select` and `index_select` but I haven't considered them yet, although I believe very little change would be needed for them. `torch_gather` is a little different, though. I think we should decompose it to `slice` for non-indexed domains and `take_along_axis`. 

I added the cross-entropy test that was originally created by @jjsjann123. It is currently segmented into two kernels, softmax as one kernel and `take_along_axis` and the final reduction as another kernel. Ideally, the `take_along_axis` op should be fused to the preceding softmax, which I'll address by a follow-up PR.

naoyam commented on csrc/index_compute.cpp:
```diff
@@ -1625,17 +1625,18 @@ std::vector<Val*> Index::getNonGlobalProducerStridedIndices(
       continue;
     }
 
+    auto override_it = override_index.find(root_dom[i]);
```
This is a bug fix. It was fine just because the non-global wasn't happening before. The override check should be done before the assertion at line 1631.

naoyam commented on csrc/ir_utils.h:
```diff
@@ -380,6 +380,14 @@ TORCH_CUDA_CU_API bool isIndexedConsumerID(
     const TensorView* tv,
     const IterDomain* id);
 
+// Return a producer ID, if any, that is indirectly accessed by, e.g.,
```
These are just really utility functions

zasdfgbnm commented on test/test_gather.cpp:
```diff
@@ -555,4 +584,399 @@ TEST_F(IndexingOpTest, GatherBroadcastInput_CUDA) {
+
+  // All of the tensors should have the split by 2, except for tv1.
+  for (auto tv : ir_utils::allTvs(&fusion)) {
+    if (tv == tv1) {
```
nit: use `ir_utils::allTvsExcept`

zasdfgbnm commented on test/test_gather.cpp:
```diff
@@ -555,4 +584,399 @@ TEST_F(IndexingOpTest, GatherBroadcastInput_CUDA) {
+  testValidate(&fusion, outputs, aten_inputs, {ref}, __LINE__, __FILE__);
+}
+
+TEST_F(IndexingOpTest, TakeAlongAxisCrossEntropyLoss_CUDA) {
```
Is there a link for the issue?

naoyam replied:
There're a couple of related issues, #156, #204, but not much detail in them. I'll update #204.

zasdfgbnm commented on test/test_gather.cpp:
```diff
@@ -555,4 +584,399 @@ TEST_F(IndexingOpTest, GatherBroadcastInput_CUDA) {
+  auto outputs = fec.runFusionWithInputs(aten_inputs);
+
+  // The reduction patterns of the normalization and the final
+  // reduction are different, so they are segmented out
```
If it is the same pattern, will they be fused? It would be nice to have a test for it.

naoyam replied:
I don't remember. It's actually not specific to `take_along_axis`, but rather it's if the persistent scheduler can fuse the second reduction. I'll add a test for it.

naoyam replied:
It does work. Added `TakeAlongAxisIntermediateTensorNormalizationAndReduction`

zasdfgbnm replied:
Great! Very happy to see this type of fusion.

zasdfgbnm commented on test/test_gather.cpp:
```diff
@@ -555,4 +584,399 @@ TEST_F(IndexingOpTest, GatherBroadcastInput_CUDA) {
+// take_along_dim to broadcast, squeeze, then normalization. Segmented
+// as the input dim to take_along_dim cannot be scheduled by the
+// reduction tv
+TEST_F(IndexingOpTest, TakeAlongAxisIntermediateTensorNormalization2_CUDA) {
```
Should we also have a test for `take_along_axis` then normalization without this issue of squeeze and not segmented?

naoyam replied:
Added `TakeAlongAxisIntermediateTensorNormalization3`

zasdfgbnm commented on test/test_gather.cpp:
```diff
@@ -555,4 +584,399 @@ TEST_F(IndexingOpTest, GatherBroadcastInput_CUDA) {
+
+// take_along_axis to broadcast, squeeze, then reduction. Segmented
+// before the reduction
+TEST_F(IndexingOpTest, TakeAlongAxisIntermediateTensorReduction2_CUDA) {
```
Should we also have a test of `take_along_axis` then reduction but without the problem of squeeze and is not segmented?

naoyam replied:
Added. `TakeAlongAxisIntermediateTensorReduction3`

zasdfgbnm commented on test/test_gather.cpp:
```diff
@@ -555,4 +584,399 @@ TEST_F(IndexingOpTest, GatherBroadcastInput_CUDA) {
+  testValidate(&fusion, outputs, aten_inputs, {ref}, __LINE__, __FILE__);
+}
+
+// Same as the above but with the pointwise scheduler
```
Looks like transpose scheduler is also support `take_along_axis`, could you add a test for it?

zasdfgbnm reviewed (commented):
Halfway through it, posting my comments so far.

naoyam commented:
@zasdfgbnm I added a few transpose tests. https://github.com/NVIDIA/Fuser/pull/250/commits/e189d44f456340bf9dba87e58257da28c6c44f42

I found a couple of issues with the Transpose scheduler, in particular the `DomainMap` in `transpose.cpp`. The tests generate correct results, but some of them fail to use the Transpose scheduler, although it should be possible to use it. I added a comment in `transpose.cpp`, but fixing it would be probably better done with the new index map, so I decided to leave it as is for now.

I'm rerunning all the tests now just in case, but nothing failed previously before the final commit. All the benchmarks also generated the same CUDA files, except for some of the `index_select` benchmarks where some variables got different names.

naoyam commented on csrc/scheduler/utils.cpp:
```diff
@@ -997,9 +997,8 @@ std::vector<TensorView*> cacheInputs(Fusion* fusion, bool unroll) {
-        ir_utils::isTorchGatherLookupTv(tv) || ir_utils::isSelectInput(tv) ||
-        ir_utils::isIndexSelectLookupTv(tv)) {
+    if (tv->uses().empty() || ir_utils::isTorchGatherLookupTv(tv) ||
+        ir_utils::isSelectInput(tv) || ir_utils::isIndexSelectLookupTv(tv)) {
```
Removed `isTorchGatherIndicesTv` as it shouldn't be necessary

zasdfgbnm replied:
`isTorchGatherIndicesTv` is no longer used anymore, should we remove it? I can not think of a case where the the index tv is special.

naoyam replied:
Removed

naoyam commented on csrc/scheduler/utils.cpp:
```diff
@@ -1357,7 +1356,6 @@ std::vector<TensorView*> getInputsOutputsWithInnerDim(
     // for index_select(lookup_tv, dim, index_tv) op
     // ignore it's lookup_tv.
     if (ir_utils::isTorchGatherLookupTv(input_tv) ||
-        ir_utils::isTorchGatherIndicesTv(input_tv) ||
```
Removed `isTorchGatherIndicesTv` as it shouldn't be necessary

naoyam commented:
Everything is green.

zasdfgbnm commented on csrc/scheduler/pointwise_utils.cpp:
```diff
@@ -6,27 +6,109 @@
+  return indexed_ids;
+}
+
+// Check if a root ID of a fusion input tensor that is indirectly
+// accessed by ops such as torch_gather needs to be mapped with
+// a reference tensor. Select has a similar effect as squeeze as the
+// indexed domain is removed, so the domain does not need to be mapped
+// as long as the tensor is a fusion input. Similarly, in index_select
+// and torch_gather, if the output domain is a broadcast, it does not
+// need to be mapped if not resolved.
```
So if I have
```c++
T2 = take_along_axis(lookup_tv=T0, index_tv=T1, axis=1)
T3 = squeeze(T2, {false, true})
```
where `T0` is `[2048, 100000000000000]`, and `T1` and `T2` are `[2048, 1]`, then can this be fused? I don't think so because we need to compute the entire `T0`, even if we just need one element of it. But the reference tensor does not have information about the second dimension of `T0`, so we will be unable to parallelize the computation.

naoyam replied:
Note that this is what's done at line 43 of the TOT version of this file.

zasdfgbnm replied:
Oh, if it is fusion input, then it's fine.

naoyam replied:
Added an assertion

zasdfgbnm commented on csrc/scheduler/pointwise_utils.cpp:
```diff
@@ -78,10 +164,34 @@ void DomainMap::eraseIfInputMappedThroughRFactorDomain(
+
+    // Further traversal if any of the new producer sets is a producer
+    // of indexed domains
+    for (const auto& exact_set : producer_sets) {
+      auto indexed_it = std::find_if(
+          indexed_sets.begin(),
+          indexed_sets.end(),
+          [&](const auto& producer_and_consumer) {
+            return producer_and_consumer.second == exact_set;
+          });
+      if (indexed_it != indexed_sets.end()) {
+        current_sets.pushBack(indexed_it->first);
+      }
+    }
```
Should we make `indexed_sets` an `unordered_map` for easier lookup?

naoyam replied:
Done

zasdfgbnm commented on csrc/scheduler/pointwise_utils.cpp:
```diff
@@ -78,10 +164,34 @@ void DomainMap::eraseIfInputMappedThroughRFactorDomain(
+  // producer-consumer pairs
+  for (auto current_sets = exact_sets; !current_sets.empty();) {
+    auto producer_sets = ca_map_.getAllDisjointSetProducers(current_sets);
+    all_exact_sets_covered.pushBack(producer_sets);
```
Should we push back the entire `producer_sets`, or should we just push back items where the check below `indexed_it != indexed_sets.end()` goes `false`?

naoyam replied:
I think pushing back the entire sets is fine. The sets in `current_sets` are those that are not yet traversed, so `producer_sets` should also be new sets.

The loop below is to traverse across indexed domains. We need this since there's nothing connecting them. I debated myself whether we should have a new IterDomain expr, or maybe we could use another ID map that allows mapping of such domains. I'm not really sure what would be the best design at this moment, so this function is making this traversal in an ad-hoc way. This may be another place where the new ID map might be helpful.

zasdfgbnm replied:
Yeah, I was also thinking about a new ID map. I think it is fine for this PR, in the future, if we see this pattern repeat, then we should add this new ID map type.

zasdfgbnm commented:
Finished reviewing, no real blockers, the only concern is https://github.com/NVIDIA/Fuser/pull/250#discussion_r1185559107 which says we are being too conservative. But being too conservative is not the end of the world. We might miss some fusion opportunities, but at least we will do everything functionally correct. So let me know if you want to leave this as-is.

naoyam commented on csrc/scheduler/registry.cpp:
```diff
@@ -435,6 +439,65 @@ class SchedulerTopologyChecker {
+      return false;
+    }
+
+    // Check if the broadcast domains are resolved. If any of them is
+    // not resolved, the reduction scheduler can't schedule the fusion
+    ConcretizedBroadcastDomains concretized_broadcast_domains(fusion);
+    if (std::any_of(
+            broadcast_consumer_of_indexed_ids.begin(),
+            broadcast_consumer_of_indexed_ids.end(),
+            [&concretized_broadcast_domains](auto id) {
+              return !concretized_broadcast_domains.isConcretized(id);
+            })) {
+      return true;
+    }
```
That makes sense. Thanks. Yes, if all domains have some connections with the reference tensor, we should be fine. Let me see if I can update the check.

zasdfgbnm commented on test/test_gather.cpp:
```diff
@@ -826,6 +826,50 @@ TEST_F(IndexingOpTest, TakeAlongAxisIntermediateTensorReduction3_CUDA) {
+  auto tv2 = add(tv0, IrBuilder::create<Double>(1));
+  auto tv3 = broadcast(tv1, {false, true});
+  auto tv4 = take_along_axis(tv2, tv3, 1);
+  auto tv5 = sum(tv4, {0, 1});
```
IIRC, `sum` will be transformed into `squeeze` then `sum`, but what we want to test is `sum` then `squeeze` is not segmented.

naoyam replied:
Yes, I realized I pushed this commit, and now testing an updated test, but I may have found a bug...

naoyam replied:
Updated `TakeAlongAxisIntermediateTensorReduction4`. It's not segmented.

Found one (likely) predication bug, which is currently worked around by inserting a `set` op. I'm pretty sure it's a generic bug, not specific to this PR. Will work on it later.

naoyam replied:
Filed an issue #282.

zasdfgbnm commented on csrc/scheduler/registry.cpp:
```diff
@@ -435,6 +439,78 @@ class SchedulerTopologyChecker {
+  // squeezed and its corresponding index-accessed producer ID, and
+  // any IDs that the producer ID depends on.
+  //
+  // This analysis has some similarity as DomainMap. Can be
+  // consolidated?
```
It don't have to be this PR, but I do think at some point we would need to cleanup this and reuse the `DomainMap` from pointwise scheduler. Also, we should cache the `DomainMap`.

naoyam replied:
I think all of these analyses should be replaced with the new `ComputeAtMap` as it should provide these traversal APIs.

naoyam commented:
!build

rdspring1 commented on test/test_gather.cpp:
```diff
@@ -555,4 +584,683 @@ TEST_F(IndexingOpTest, GatherBroadcastInput_CUDA) {
+
+// Similar to TakeAlongAxisIntermediateTensorReduction2, but no
+// squeeze of the consumer ID of the indexed domain. Should not be segmented.
+TEST_F(IndexingOpTest, TakeAlongAxisIntermediateTensorReduction4_CUDA) {
```
I am seeing this error with this test on my 3090: 

```
[ RUN      ] IndexingOpTest.TakeAlongAxisIntermediateTensorReduction4_CUDA
error: CUDA_ERROR_COOPERATIVE_LAUNCH_TOO_LARGE failed with error too many blocks in cooperative launch
```

naoyam replied:
Oops, sorry. I'll update the test.

naoyam replied:
This should fix the error https://github.com/NVIDIA/Fuser/pull/292. Only 40 blocks are launched.

zasdfgbnm replied:
Cross referencing https://github.com/NVIDIA/Fuser/issues/284
